src/utl/src/ScopedTemporaryFile.cpp

```diff
@@ -33,4 +36,61 @@ ScopedTemporaryFile::~ScopedTemporaryFile()
   }
 }
 
+StreamHandler::StreamHandler(const char* filename,
+                             std::ios_base::iostate flag,
+                             std::ios_base::openmode mode)
+    : filename_(filename), flag_(flag), mode_(mode)
+{
+  if (fs::exists(filename_)) {
+    fs::remove(filename_);
+  }
+  std::string tmp_filename = filename_ + ".tmp";
+  if (fs::exists(tmp_filename)) {
+    fs::remove(tmp_filename);
+  }
+  os_.exceptions(flag | std::ofstream::failbit);
+  os_.open(tmp_filename, mode);
+}
+
+StreamHandler::~StreamHandler()
+{
+  if (os_.is_open()) {
+    os_.close();
+  }
+  std::string tmp_filename = filename_ + ".tmp";
+  fs::rename(tmp_filename, filename_);
+}
+
+std::ofstream& StreamHandler::getStream()
+{
+  return os_;
+}
+
+FileHandler::FileHandler(const char* filename, const char* mode)
+    : filename_(filename), mode_(mode)
+{
+  if (fs::exists(filename_)) {
+    fs::remove(filename_);
+  }
+  std::string tmp_filename = filename_ + ".tmp";
+  file = fopen(tmp_filename.c_str(), mode);
+  if (!file) {
+    throw std::runtime_error("Failed to open temporary file");
```

Comment:
read errno and elaborate the exception message a bit with filename that could not be opened.

Search code for errno to find example.